easypost/webhook.py

```diff
@@ -17,3 +28,36 @@ def update(self, **params) -> "Webhook":
         response, api_key = requestor.request(method=RequestMethod.PATCH, url=url, params=params)
         self.refresh_from(values=response, api_key=api_key)
         return self
+
+    @staticmethod
+    def validate_webhook(event_body: bytes, headers: Dict[str, Any], webhook_secret: str) -> Dict[str, Any]:
+        """Validate a webhook by comparing the HMAC signature header sent from EasyPost to your shared secret.
+        If the signatures do not match, an error will be raised signifying the webhook either did not originate
+        from EasyPost or the secrets do not match. If the signatures do match, the `event_body` will be returned
+        as JSON.
+        """
+        easypost_hmac_signature = headers.get("X-Hmac-Signature")
+
+        if easypost_hmac_signature:
+            normalized_secret = unicodedata.normalize("NFKD", webhook_secret)
+            encoded_secret = bytes(normalized_secret, "utf8")
+            hexified_secret = binascii.hexlify(encoded_secret)
+
+            expected_signature = hmac.new(
+                key=hexified_secret,
+                msg=event_body,
+                digestmod=hashlib.sha256,
+            )
+
+            digest = "hmac-sha256-hex=" + expected_signature.hexdigest()
+
+            if hmac.compare_digest(digest, easypost_hmac_signature):
+                webhook_body = json.loads(event_body)
+            else:
+                raise Error(
+                    message="Webhook received did not originate from EasyPost or had a webhook secret mismatch."
+                )
+        else:
+            raise Error(message="Webhook received does not contain an HMAC signature.")
+
+        return webhook_body
```

Comment:
Still not a fan that this function does two separate things, but I won't block it.

Reply:
It validates your webhook then gives it to you if there is no error, it's not doing two things. You send your webhook in and get it out barring no error.

Reply:
> It validates your webhook then gives it to you if there is no error

Exactly, it's doing one thing (validating it) and then doing another thing (deserializing the JSON; what if the webhook is ever not JSON?)

Deserializing the webhook should be done in a separate, explicit step by the end-user.

Reply:
This decision was made from feedback on an internal discussion:

> what james is saying that the body goes in as a series of bytes, and what comes out is a deserialized structure; you want to build the signature validation into the routine that deserializes. If the signature validation fails, it raises an exception. This way you don't end up with a state where you have a deserialized structure that is the result of an invalid signature